backport std::ssize to c10

dagitses opened the pull request:
backport std::ssize to c10

Summary:
Now that we have -Werror=sign-compare enabled, we encounter a lot of
friction comparing standard containers and our tensors which are
signed.

std::ssize will make it easier and safer to succinctly convert
container sizes to a signed type.

Test Plan: Added a unit test.

Reviewers: ezyang

Subscribers:

Tasks:

Tags:

---
Stack created with [Sapling](https://sapling-scm.com). Best reviewed with [ReviewStack](https://reviewstack.dev/pytorch/pytorch/pull/97442).
* #97443
* __->__ #97442

ezyang commented on c10/util/ssize.h:
```diff
@@ -0,0 +1,46 @@
+  // We use an assert here because we expect this to be exceedingly
+  // rare to fire and don't wish to pay a performance hit in release
+  // mode.
+  assert(!greater_than_max<R>(c.size()));
```
you can use `TORCH_INTERNAL_ASSERT_DEBUG_ONLY`

dagitses replied:
done!

Commit:
backport std::ssize to c10

Summary:
Now that we have -Werror=sign-compare enabled, we encounter a lot of
friction comparing standard containers and our tensors which are
signed.

std::ssize will make it easier and safer to succinctly convert
container sizes to a signed type.

Test Plan: Added a unit test.

Reviewers: ezyang

Subscribers:

Tasks:

Tags:

dagitses commented:
@pytorchbot merge
